### ISSUE TITLE
CTR in statusBlob has different bit size than in calculations

### ISSUE THREAD

**Issue by hvge** [lime-company/lime-security-powerauth#51]
The CTR returned in statusBlob is 32 bit value but every operation related to CTR uses 64 bit counter. I know that the CTR is just an informational value here and right now we don't have practical usage for it. But still I think that this should be adjusted and we may use one common type for the counter.

You'll probably need to extend the blob to two AES blocks to keep enough entropy.

Btw, if you gonna to accept this issue and extend the structure, you should use different "magic" value. The 0xDEADBEEF is so lame, because everybody is using it. What about 0xDEC0DEAD / 0xDECODEED or anything more funny (https://en.wikipedia.org/wiki/Hexspeak:) 

...in fact, I created this issue just because that "magic" :) We've been already discussing this topic personally and I know that you'll extend the blob someday.

**Comment by hvge** [comment]
Oh, I have a better idea! 
What about to use just one byte. Basically, we don't need that counter at all. What we need is to give client chance to determine, if server is completely of out sync or not. Check this out:

1. Let call that byte as CTRH (counter hint) = CTR & 0xFF
2. Server sends its CTRHs in status blob as an one byte
3. Client calculates his local CTRHc
4. Let calculate the difference:```
	int calculateDifference(byte CTRHs, byte CTRHc)
	{
	    int diff = (int)CTRHs - (int)CTRHc;
	    if (diff > 127) {
	        // CTRHs is too big. The better explanation for
	        // this situation is that counter on client's side
	        // overflowed over its lowest byte and client is
	        // really ahead against the server.
	        diff -= 256;
	    } else if (diff < -128) {
	        // CTRHs is too small. The better explanation for
	        // the situation is that counter on server's side
	        // overflowed over its lowest byte and server
	        // is really ahead against the client.
	        diff += 256;
	    }   
			// ...in all other situations, the difference
	        // looks OK and in the limits.
	    return diff;
	}
```
5. If difference is greater or equal to 0, then the client is ahead and that's great.
   We can compare that value against tolerance and decide what to do next.

6. If difference is less than 0, then the server is ahead a we'll never have a chance
   to calculate a valid signature. 

That should work pretty accurate, if the tolerance on server's side is way below 127

**Comment by petr-inmite** [comment]
I don't think the 2nd solution would work well, since the numbers are cyclic. If you receive 41 from server, and you compute 62, it may mean that you are 21 ahead the server, or that server is way ahead of you. In other words, we are assuming it will somehow work since the counters shouldn't be too much out of sync.

I will extend the status blob to have 32 bytes instead of 16, the change will be available in 0.6.

Good point with the "magic" - let's avoid anything funny and keep it serious. `0xDEC0DED1` (like "decoded = true") would be a good choice, IMO.

**Comment by petr-inmite** [comment]
Current proposed status:

Structure of the 32B long status blob is following:

```java
	0xDE 0xAD 0xBE 0xEF 1B:${STATUS} 8B:${CTR} 1B:${FAIL_COUNT} 1B:${MAX_FAIL_COUNT} 17B:${RANDOM_NOISE}
```

where:

- The first 4 bytes (`0xDE 0xC0 0xDE 0xD1`) are basically a fixed prefix.
- `${STATUS}` - A status of the activation record, it can be one of following values:
	- `0x01 - CREATED`
	- `0x02 - OTP_USED`
	- `0x03 - ACTIVE`
	- `0x04 - BLOCKED`
	- `0x05 - REMOVED`
- `${CTR}` - 8 bytes representing information of the server counter (`CTR` value, as defined in PowerAuth 2.0 specification).
- `${FAIL_COUNT}` - 1 byte representing information about the number of failed attempts at the moment.
- `${MAX_FAIL_COUNT}` - 1 byte representing information about the maximum allowed number of failed attempts.
- `${RANDOM_NOISE}` - Random 17 byte padding (a complement to the total length of 32B). These bytes also serve as a source of entropy for the transport (AES encrypted `cStatusBlob` will be different each time an endpoint is called).
